compiler/rustc_codegen_llvm/src/type_of.rs

```diff
@@ -443,8 +443,36 @@ impl<'tcx> LayoutLlvmExt<'tcx> for TyAndLayout<'tcx> {
             return Some(cx.type_vector(ety, *count));
         }
 
-        // FIXME: The above only handled integer arrays; surely more things
-        // would also be possible. Be careful about provenance, though!
-        None
+        // Ensure the type isn't too complex nor otherwise ineligible
+        is_scalar_copy_reasonable(4, self.ty, cx)?;
+
+        // But otherwise we can load/store the LLVM type directly
+        Some(self.llvm_type(cx))
+    }
+}
+
+fn is_scalar_copy_reasonable<'a, 'tcx>(
+    max_fields: u32,
+    t: Ty<'tcx>,
+    cx: &CodegenCx<'a, 'tcx>,
+) -> Option<u32> {
+    if t.is_any_ptr() || t.is_primitive() {
+        return max_fields.checked_sub(1);
+    }
```

Comment:
Do ZSTs still exist at this point? if so then it would make sense to not count them.

Reply:
They do, because this is looking at the Rust type model, but this is looking recursively, so ZSTs end up not counting because they has no fields: `()` is an empty `ty::Tuple`, for example, and `PhantomData` is a unit struct that also has no fields.